### PR TITLE
send empty messages on last chat response

### PR DESCRIPTION
**Pull request by BruceMacD**
Send an empty message on the last chat response rather than omitting it. This makes the chat API match the generate API.

As of this change...

`/generate`
```
curl http://localhost:11434/api/generate -d '{
    "model": "mistral"
}'

HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Date: Thu, 14 Dec 2023 22:42:13 GMT
Content-Length: 88
Connection: close

{
  "model": "mistral",
  "created_at": "2023-12-14T22:42:13.365069Z",
  "response": "",
  "done": true
}
```
```
curl http://localhost:11434/api/generate -d '{
    "model": "mistral",
    "prompt": "reply with a single word"
}'

HTTP/1.1 200 OK
Content-Type: application/x-ndjson
Date: Thu, 14 Dec 2023 22:43:05 GMT
Connection: close
Transfer-Encoding: chunked

{"model":"mistral","created_at":"2023-12-14T22:43:05.60168Z","response":" Okay","done":false}
{"model":"mistral","created_at":"2023-12-14T22:43:05.616406Z","response":".","done":false}
{"model":"mistral","created_at":"2023-12-14T22:43:05.631163Z","response":"","done":true,"context":[733,16289,28793,28705,10071,395,264,2692,1707,733,28748,16289,28793,19811,28723],"total_duration":414743625,"load_duration":760500,"prompt_eval_count":14,"prompt_eval_duration":393924000,"eval_count":2,"eval_duration":14727000}
```

`/chat`
```
curl http://localhost:11434/api/chat -d '{
    "model": "mistral"
}'

HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Date: Thu, 14 Dec 2023 22:41:56 GMT
Content-Length: 132
Connection: close

{
  "model": "mistral",
  "created_at": "2023-12-14T22:41:56.540246Z",
  "message": {
    "role": "assistant",
    "content": "",
    "images": null
  },
  "done": true
}
```
```
curl http://localhost:11434/api/chat -d '{
    "model": "mistral",
    "messages": [
        {
            "role": "user",
            "content": "reply with one word"
        }
    ]
}'
HTTP/1.1 200 OK
Content-Type: application/x-ndjson
Date: Thu, 14 Dec 2023 22:38:36 GMT
Connection: close
Transfer-Encoding: chunked

{"model":"mistral","created_at":"2023-12-14T22:38:36.625168Z","message":{"role":"assistant","content":" Okay","images":null},"done":false}
{"model":"mistral","created_at":"2023-12-14T22:38:36.639622Z","message":{"role":"assistant","content":".","images":null},"done":false}
{"model":"mistral","created_at":"2023-12-14T22:38:36.654161Z","message":{"role":"assistant","content":"","images":null},"done":true,"total_duration":417861167,"load_duration":927584,"prompt_eval_count":13,"prompt_eval_duration":396099000,"eval_count":2,"eval_duration":14502000}
```

### CODE REVIEW

File: api/types.go [hunk 1]
```diff
@@ -59,13 +59,13 @@ type ChatRequest struct {
 type Message struct {
 	Role    string      `json:"role"` // one of ["system", "user", "assistant"]
 	Content string      `json:"content"`
-	Images  []ImageData `json:"images, omitempty"`
```
Comment (BruceMacD) [inline api/types.go:62]: the space here before omitempty means this was never omitted, so removing this does not change behavior from the current API

Comment (pdevine) [reply api/types.go:62]: should we return it if it's empty though?

Comment (BruceMacD) [reply api/types.go:62]: To match the behavior of other parts of the message I say yes, but its not very aesthetically pleasing.

File: api/types.go [hunk 2]
```diff
@@ -59,13 +59,13 @@ type ChatRequest struct {
...
+	Images  []ImageData `json:"images"`
 }
 
 type ChatResponse struct {
 	Model     string    `json:"model"`
 	CreatedAt time.Time `json:"created_at"`
-	Message   *Message  `json:"message,omitempty"`
+	Message   *Message  `json:"message"`
```
Comment (jmorganca) [inline api/types.go:68]: Should this be a pointer still?

File: api/types.go (hunk 1 above)
Comment (jmorganca) [reply api/types.go:62]: It shouldn't (for now), since this would imply the response _might_ contain an image, which it won't!

Comment (jmorganca) [reply api/types.go:62]: E.g. today we get this

```
"message":{"role":"assistant","content":" right","images":null}
```

it should be

```
"message":{"role":"assistant","content":" right"}
```

**Review by jmorganca — APPROVED** [review]

**Pull request merged**